weaveworks/flintlock
# feat: ability to list microvms based on name

**What this PR does / why we need it**:

### Background

I considered two routes:

1. Create a new Find() method
2. Add the name field on the List() method

The second option sounds better because it's already a List method, so
makes sense to list mirovms that way.

The extra name field is optional, if it's not provided it works the same
way as before, but if it's provided it will filter based on namespace
and name.

### Changes

#### proto and gRPC server

Changing the definition to use name in List queries.

#### tests

* Update all unit tests.
* Update e2e tests.
* Add extra case to e2e test: List request with name field and expect
  only one MicroVM in the response.

#### GetAll function chain

In the background I change the whole GetAll function chain to use a
filter `map[string]string`, that way we can use them to filter based on
different filters.

I considered doing a much cleaner solution something like this:

    // WithNamespace is a namespace filter for queries.
    func WithNamespace(namespace string) models.ListMicroVMQuery {
           return func() (string, string) {
                   return NamespaceLabel(), namespace
           }
    }

    // WithName is a name filter for queries.
    func WithName(name string) models.ListMicroVMQuery {
           return func() (string, string) {
                   return NameLabel(), namespace
           }
    }

But at the end, I couldn't find a good place for it becasue that should
live under containerd as those labels are containerd specific formats,
but then I couldn't think of an easy way to astract it through ports and
creating an outer layer for filters and put it in ports seems
unnecessary and overkill, so I stuck with the simple key-value map.

#### Fix list results

List endpoints returned only one of the MicroVMs when we have multiple MicroVMs with the same ns/name.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #368
Fixes #370

**Special notes for your reviewer**:
